### ISSUE TITLE
Using PancakeView with Entry in TitleView throws NotSupportedException on Android

### ISSUE THREAD

**Issue by acuntex** [sthewissen/Xamarin.Forms.PancakeView#113]
If you use PancakeView with an Entry in NavigationPage.TitleView, it throws a NotSupportedException if you touch the entry and start to type.

I've also seen the same exception in a PancakeView with an Entry on the Page itself, but this seems to be very rare, couldn't reproduce it anymore.

The following code was used:
```
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:d="http://xamarin.com/schemas/2014/forms/design"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:pancakeView="clr-namespace:Xamarin.Forms.PancakeView;assembly=Xamarin.Forms.PancakeView"
             mc:Ignorable="d"
             x:Class="PancakeViewBugTitleView.MainPage">
    <NavigationPage.TitleView>
        <StackLayout>
            <pancakeView:PancakeView HeightRequest="40" CornerRadius="20,20,20,20" BackgroundColor="Gray">
                <pancakeView:PancakeView.Border>
                    <pancakeView:Border Color="#33333c" Thickness="1" />
                </pancakeView:PancakeView.Border>
                <StackLayout Orientation="Horizontal" HeightRequest="20" Padding="10,0,10,0">
                    <Image Source="icon_search.png" HorizontalOptions="Start" VerticalOptions="CenterAndExpand"/>
                    <Entry Placeholder="Click here, try to type, it will crash"
                                          HorizontalOptions="FillAndExpand"
                                          VerticalOptions="CenterAndExpand"
                                          MaxLength="20"
                                          IsTextPredictionEnabled="False"
                                          IsSpellCheckEnabled="False"
                                          ReturnType="Done">
                    </Entry>
                </StackLayout>
            </pancakeView:PancakeView>
        </StackLayout>
    </NavigationPage.TitleView>
    <ContentPage.Content>
        <StackLayout>
            <pancakeView:PancakeView HeightRequest="40" CornerRadius="20,20,20,20" BackgroundColor="Gray">
                <pancakeView:PancakeView.Border>
                    <pancakeView:Border Color="#33333c" Thickness="1" />
                </pancakeView:PancakeView.Border>
                <StackLayout Orientation="Horizontal" HeightRequest="20" Padding="10,0,10,0">
                    <Image Source="icon_search.png" HorizontalOptions="Start" VerticalOptions="CenterAndExpand"/>
                    <Entry Placeholder="Sometimes crashes here too, but it's very rare"
                                          HorizontalOptions="FillAndExpand"
                                          VerticalOptions="CenterAndExpand"
                                          MaxLength="20"
                                          IsTextPredictionEnabled="False"
                                          IsSpellCheckEnabled="False"
                                          ReturnType="Done">
                    </Entry>
                </StackLayout>
            </pancakeView:PancakeView>
        </StackLayout>
    </ContentPage.Content>
</ContentPage>

```
Here is the Exception:
```
07-14 13:46:32.007 I/MonoDroid(29578): UNHANDLED EXCEPTION:
07-14 13:46:32.043 I/MonoDroid(29578): System.NotSupportedException: Unable to activate instance of type Xamarin.Forms.PancakeView.Droid.PancakeViewRenderer from native handle 0x61 (key_handle 0x9ebaae5). ---> System.MissingMethodException: No constructor found for Xamarin.Forms.PancakeView.Droid.PancakeViewRenderer::.ctor(System.IntPtr, Android.Runtime.JniHandleOwnership) ---> Java.Interop.JavaLocationException: Exception of type 'Java.Interop.JavaLocationException' was thrown.
07-14 13:46:32.043 I/MonoDroid(29578):    --- End of inner exception stack trace ---
07-14 13:46:32.043 I/MonoDroid(29578):   at Java.Interop.TypeManager.CreateProxy (System.Type type, System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x000b5] in <55654ebe9f2a48e6bade2862bb243f94>:0 
07-14 13:46:32.043 I/MonoDroid(29578):   at Java.Interop.TypeManager.CreateInstance (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type targetType) [0x00111] in <55654ebe9f2a48e6bade2862bb243f94>:0 
07-14 13:46:32.043 I/MonoDroid(29578):    --- End of inner exception stack trace ---
07-14 13:46:32.043 I/MonoDroid(29578):   at Java.Interop.TypeManager.CreateInstance (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type targetType) [0x0017e] in <55654ebe9f2a48e6bade2862bb243f94>:0 
07-14 13:46:32.043 I/MonoDroid(29578):   at Java.Lang.Object.GetObject (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type type) [0x00023] in <55654ebe9f2a48e6bade2862bb243f94>:0 
07-14 13:46:32.043 I/MonoDroid(29578):   at Java.Lang.Object._GetObject[T] (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00017] in <55654ebe9f2a48e6bade2862bb243f94>:0 
07-14 13:46:32.043 I/MonoDroid(29578):   at Java.Lang.Object.GetObject[T] (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00000] in <55654ebe9f2a48e6bade2862bb243f94>:0 
07-14 13:46:32.043 I/MonoDroid(29578):   at Android.Views.View.get_Parent () [0x0001f] in <55654ebe9f2a48e6bade2862bb243f94>:0 
07-14 13:46:32.043 I/MonoDroid(29578):   at Xamarin.Forms.Platform.Android.DescendantFocusToggler.RequestFocus (Android.Views.View control, System.Func`1[TResult] baseRequestFocus) [0x0002f] in <312f0ce8dbcc4c47b5fc27a24af8b2b0>:0 
07-14 13:46:32.043 I/MonoDroid(29578):   at Xamarin.Forms.Platform.Android.FormsEditTextBase.Xamarin.Forms.Platform.Android.IDescendantFocusToggler.RequestFocus (Android.Views.View control, System.Func`1[TResult] baseRequestFocus) [0x00015] in <312f0ce8dbcc4c47b5fc27a24af8b2b0>:0 
07-14 13:46:32.043 I/MonoDroid(29578):   at Xamarin.Forms.Platform.Android.FormsEditTextBase.RequestFocus (Android.Views.FocusSearchDirection direction, Android.Graphics.Rect previouslyFocusedRect) [0x0001b] in <312f0ce8dbcc4c47b5fc27a24af8b2b0>:0 
07-14 13:46:32.043 I/MonoDroid(29578):   at Android.Views.View.n_RequestFocus_ILandroid_graphics_Rect_ (System.IntPtr jnienv, System.IntPtr native__this, System.Int32 native_direction, System.IntPtr native_previouslyFocusedRect) [0x00011] in <55654ebe9f2a48e6bade2862bb243f94>:0 
07-14 13:46:32.043 I/MonoDroid(29578):   at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.76(intptr,intptr,int,intptr)
07-14 13:46:32.531 E/AndroidRuntime(29578): FATAL EXCEPTION: main
07-14 13:46:32.531 E/AndroidRuntime(29578): Process: com.app, PID: 29578
07-14 13:46:32.531 E/AndroidRuntime(29578): crc64fe7e538dcffb0c36.MonoExceptionHelper: (NotSupportedException) Unable to activate instance of type Xamarin.Forms.PancakeView.Droid.PancakeViewRenderer from native handle 0x61 (key_handle 0x9ebaae5).
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Java.Interop.TypeManager.CreateInstance(System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type targetType)[0x0017e](Java/Interop/TypeManager0x0017e.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Java.Lang.Object.GetObject(System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type type)[0x00023](Java/Lang/Object0x00023.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Java.Lang.Object._GetObject[T](System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer)[0x00017](Java/Lang/Object0x00017.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Java.Lang.Object.GetObject[T](System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer)[0x00000](Java/Lang/Object0x00000.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Android.Views.View.get_Parent()[0x0001f](Android/Views/View0x0001f.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Xamarin.Forms.Platform.Android.DescendantFocusToggler.RequestFocus(Android.Views.View control, System.Func`1[TResult] baseRequestFocus)[0x0002f](Xamarin/Forms/Platform/Android/DescendantFocusToggler0x0002f.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Xamarin.Forms.Platform.Android.FormsEditTextBase.Xamarin.Forms.Platform.Android.IDescendantFocusToggler.RequestFocus(Android.Views.View control, System.Func`1[TResult] baseRequestFocus)[0x00015](Xamarin/Forms/Platform/Android/FormsEditTextBase/Xamarin/Forms/Platform/Android/IDescendantFocusToggler0x00015.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Xamarin.Forms.Platform.Android.FormsEditTextBase.RequestFocus(Android.Views.FocusSearchDirection direction, Android.Graphics.Rect previouslyFocusedRect)[0x0001b](Xamarin/Forms/Platform/Android/FormsEditTextBase0x0001b.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Android.Views.View.n_RequestFocus_ILandroid_graphics_Rect_(System.IntPtr jnienv, System.IntPtr native__this, System.Int32 native_direction, System.IntPtr native_previouslyFocusedRect)[0x00011](Android/Views/View0x00011.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at   at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.76(intptr,intptr,int,intptr).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): Caused by: crc64fe7e538dcffb0c36.MonoExceptionHelper: (MissingMethodException) No constructor found for Xamarin.Forms.PancakeView.Droid.PancakeViewRenderer::.ctor(System.IntPtr, Android.Runtime.JniHandleOwnership)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Java.Interop.TypeManager.CreateProxy(System.Type type, System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer)[0x000b5](Java/Interop/TypeManager0x000b5.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Java.Interop.TypeManager.CreateInstance(System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type targetType)[0x00111](Java/Interop/TypeManager0x00111.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): Caused by: crc64fe7e538dcffb0c36.MonoExceptionHelper: (JavaLocationException) Exception of type 'Java.Interop.JavaLocationException' was thrown.
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at Java.Lang.Error: Exception of type 'Java.Lang.Error' was thrown..(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at   --- End of managed Java.Lang.Error stack trace ---.(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at java.lang.Error: Java callstack:.(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at crc643f46942d9dd1fff9.FormsEditTextBase.n_requestFocus(Native Method).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at crc643f46942d9dd1fff9.FormsEditTextBase.requestFocus(FormsEditTextBase.java:53).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.view.View.requestFocus(View.java:12367).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.view.View.requestFocus(View.java:12309).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at crc643f46942d9dd1fff9.FormsEditText.n_onSelectionChanged(Native Method).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at crc643f46942d9dd1fff9.FormsEditText.onSelectionChanged(FormsEditText.java:62).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.widget.TextView.spanChange(TextView.java:10684).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.widget.TextView$ChangeWatcher.onSpanChanged(TextView.java:13400).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.text.SpannableStringBuilder.sendSpanChanged(SpannableStringBuilder.java:1307).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.text.SpannableStringBuilder.sendToSpanWatchers(SpannableStringBuilder.java:651).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:580).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:507).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:37).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.view.inputmethod.BaseInputConnection.replaceText(BaseInputConnection.java:843).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.view.inputmethod.BaseInputConnection.setComposingText(BaseInputConnection.java:616).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at com.android.internal.view.IInputConnectionWrapper.executeMessage(IInputConnectionWrapper.java:393).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at com.android.internal.view.IInputConnectionWrapper$MyHandler.handleMessage(IInputConnectionWrapper.java:89).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.os.Handler.dispatchMessage(Handler.java:107).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.os.Looper.loop(Looper.java:214).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at android.app.ActivityThread.main(ActivityThread.java:7356).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at java.lang.reflect.Method.invoke(Native Method).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492).(unknown.cs:1)
07-14 13:46:32.531 E/AndroidRuntime(29578): 	at 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930).(unknown.cs:1)
```

I also created a repro repository:
https://github.com/acuntex/PancakeViewBug_TitleView

It's possible to add the missing constructor, but it's more import why he suddenly needs that weird constructor.

**Comment by sthewissen** [comment]
As per Samantha's comment here it might be the linker removing PancakeView:

https://github.com/xamarin/Xamarin.Forms/issues/11098

You could try calling `PancakeViewRenderer.Init();` from your MainActivity's `OnCreate`.

**Comment by sthewissen** [comment]
I also wonder whether this is a bug specific to PancakeView or if something similar happens with other types of views.

**Comment by acuntex** [comment]
No, PancakeView works fine when it's not in the TitleView. The problem is that somehow the Framework is looking for another constructor with two parameters (System.IntPtr, Android.Runtime.JniHandleOwnership)

The bug occurs also on Debug without any optimizations, so there's no R8 or anything active. I've seen that Xamarin had similar issues with the LabelRenderer or other Renderers before, according to them it's fixed, apparently not.

The following workaround works:

```
    public class TitleViewPancakeView: PancakeView
    {
    }

```
```
[assembly: ExportRenderer(typeof(TitleViewPancakeView), typeof(CustomPancakeViewRenderer))]
namespace YourApp.Droid.Renderer
{
    public class CustomPancakeViewRenderer: PancakeViewRenderer
    {
        public CustomPancakeViewRenderer(Context context) : base(context)
        {
        }
        
        [Obsolete]
        public CustomPancakeViewRenderer(IntPtr handle, JniHandleOwnership transfer): base(Xamarin.Forms.Forms.Context)
        {
        }
    }
}
```

Just make sure to set the Visual of the PancakeView to Default because if it is not, the DependencyResolver of Xamarin.Forms (https://github.com/xamarin/Xamarin.Forms/blob/719fc7a604ff0cce8922d717c99bfb0fa17e35e0/Xamarin.Forms.Core/DependencyResolver.cs) has a pretty dirty workaround in line 50, which checks if the constructor has two parameters (without any type checking) and then fails because it does not match.

In the Repro-Repository you can see both cases, in titleview and not in titleview, it just crashes when you start typing.

**Comment by sthewissen** [comment]
I can add a constructor to the renderer as a workaround?

```csharp
[Obsolete]
public PancakeViewRenderer(IntPtr handle, JniHandleOwnership transfer) : base(Xamarin.Forms.Forms.Context)
{
}
```

**Comment by acuntex** [comment]
It would work, but you would also have to set the Visual of the PancakeView to the Default because if a parent Visual is e.g. Material, it will crash because of the dirty workaround there is in the DependencyResolver in Xamarin.Forms.

Basically a workaround of a workaround because Xamarin uses workarounds :-(

**Comment by sthewissen** [comment]
Guess I'll try. Have to admit that a different solution doesn't quite spring to mind just yet.

**Comment by sthewissen** [comment]
I would say overriding the Visual property and not providing a setter should do the trick. Thoughts?

```
public new IVisual Visual => VisualMarker.Default;
```

**Comment by dhindrik** [comment]
I got this only when I navigate back to a page. And I think it started after I upgraded my project to AndroidX.

**Comment by acuntex** [comment]
@dhindrik The error seems to occur when Android tries to Dispose the Renderer... It's also on versions < AndroidX.

I'm starting to believe it's a Xamarin.Forms Bug but they don't want to listen. :(

**Comment by sthewissen** [comment]
Could you try this in https://www.nuget.org/packages/Xamarin.Forms.PancakeView/2.0.1.701-beta?

**Comment by acuntex** [comment]
Made a comment in the commit.

Will check it out, if you want a fast check, you can get https://github.com/acuntex/PancakeViewBug_TitleView and start it in the android Emulator.
The current version should crash, after you updated it should work hopefully.

**Comment by sthewissen** [comment]
Confirmed fixed in #116
